### PR TITLE
Add tree_pos to tsk_tree_t struct

### PR DESCRIPTION
**Pull request by duncanMR**
## Description

This PR aims to address the first part of #2795 by adding the `tree_position` class to the `tsk_tree_t` struct. I've done a first draft of this and it passes all the unit tests without any memory issues detected by Valgrind. I'm not confident that the memory management is correct yet, though!

# PR Checklist:

- [ ] Tests that fully cover new/changed functionality.
- [ ] Documentation including tutorial content if appropriate.
- [ ] Changelogs, if there are API changes.

### CODE REVIEW

**Comment by duncanMR** [comment]
@jeromekelleher I see why you made the comment about doxygen in the trees header file: moving `tsk_position_t` above `tsk_tree_t` broke the docs!

File: c/tskit/trees.c [hunk 1]
```diff
@@ -4613,6 +4621,7 @@ tsk_tree_free(tsk_tree_t *self)
     tsk_safe_free(self->next_sample);
     tsk_safe_free(self->num_children);
     tsk_safe_free(self->edge);
+    tsk_safe_free(self->tree_pos);
```
Comment (jeromekelleher) [inline c/tskit/trees.c:4624]: This is wrong , and I'm surprised valgrind doesn't catch it. The golden rule is, you only ``free`` a pointer than you have previously ``malloc``ed. The ``tree_pos`` here is an embedded struct, so it's not  a pointer that can be freed.

You should be calling ``tsk_tree_position_free(&self->tree_pos);`` instead.

File: c/tskit/trees.h [hunk 2]
```diff
@@ -111,6 +111,37 @@ typedef struct {
...
+        tsk_id_t stop;
+        const tsk_id_t *order;
+    } out;
+    tsk_id_t left_current_index;
+    tsk_id_t right_current_index;
+    int direction;
+    const tsk_treeseq_t *tree_sequence;
+} tsk_tree_position_t;
+
+int tsk_tree_position_init(
```
Comment (jeromekelleher) [inline c/tskit/trees.h:136]: Methods are declared further down the file, separate from struct definitions. Just move the function signatures to the bottom of the file.

**Comment by duncanMR** [comment]
@jeromekelleher, I've added a commit with all your suggested changes. The docs are still failing after moving all the methods to the bottom of `trees.h`. Doxygen runs successfully, and it seems like the xml file for the `tsk_tree_t` class it generates is correct. Also, the tree section of `docs/_build/html/c-api.html` looks correct to me; I'm confused as to what is causing [the warning](https://github.com/tskit-dev/tskit/actions/runs/6767668792/job/18390667443?pr=2867).

**Comment by jeromekelleher** [comment]
I think the docs build is working @duncanMR - I just checked it out and the build is fine for me. Also the CI docs build is working.

I think the CI failures are due to a stale package cache. Probably simplest to just start a new PR from a fresh branch.

**Comment by duncanMR** [comment]
> I think the docs build is working @duncanMR - I just checked it out and the build is fine for me. Also the CI docs build is working.
> 
> I think the CI failures are due to a stale package cache. Probably simplest to just start a new PR from a fresh branch.

Yes, I figured out the docs problem: there is a hardcoded exception in the Sphinx config specifically for the tree class that needed an update. I'll make a new PR now

**Comment by jeromekelleher** [comment]
I just saw a "merge" here @duncanMR - merge is not your friend! Rebase rebase rebase

**Comment by duncanMR** [comment]
@jeromekelleher Sorry about that! I accidentally pulled changes instead of just fetching. Will be more careful in future

**Pull request closed without merging**